File: backend/internal/handlers/contacts.go
package handlers

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strconv"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"crm-api/internal/database"
	"crm-api/internal/middleware"
	"crm-api/internal/scoring"
)

type Contact struct {
	ID                string          `json:"id"`
	AgentID           string          `json:"agent_id"`
	FirstName         string          `json:"first_name"`
	LastName          string          `json:"last_name"`
	Email             *string         `json:"email"`
	Phone             *string         `json:"phone"`
	Source            *string         `json:"source"`
	FolderID          *string         `json:"folder_id"`
	FolderName        *string         `json:"folder_name"`
	LeadScore         int             `json:"lead_score"`
	LeadScoreSignals  json.RawMessage `json:"lead_score_signals"`
	PreviousLeadScore *int            `json:"previous_lead_score"`
	CreatedAt         time.Time       `json:"created_at"`
	UpdatedAt         time.Time       `json:"updated_at"`
}

func ListContacts(pool *pgxpool.Pool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		agentID := middleware.AgentUUIDFromContext(r.Context())
		if agentID == "" {
			respondErrorWithCode(w, http.StatusUnauthorized, "unauthorized", ErrCodeUnauthorized)
			return
		}

		search := r.URL.Query().Get("search")
		source := r.URL.Query().Get("source")
		folderFilter := r.URL.Query().Get("folder_id")
		sortParam := r.URL.Query().Get("sort")
		page, _ := strconv.Atoi(r.URL.Query().Get("page"))
		limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
		if page < 1 {
			page = 1
		}
		if limit < 1 || limit > 100 {
			limit = 25
		}
		offset := (page - 1) * limit

		tx, err := database.BeginWithRLS(r.Context(), pool, agentID)
		if err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "database error", ErrCodeDatabase)
			return
		}
		defer tx.Rollback(r.Context())

		var args []interface{}
		whereExpr := "1=1"

		if search != "" {
			args = append(args, "%"+search+"%")
			n := len(args)
			whereExpr += fmt.Sprintf(" AND (c.first_name ILIKE $%d OR c.last_name ILIKE $%d OR c.email ILIKE $%d)", n, n, n)
		}
		if source != "" {
			args = append(args, source)
			whereExpr += fmt.Sprintf(" AND c.source = $%d", len(args))
		}
		if folderFilter == "unfiled" {
			whereExpr += " AND c.folder_id IS NULL"
		} else if folderFilter != "" {
			args = append(args, folderFilter)
			whereExpr += fmt.Sprintf(" AND c.folder_id = $%d", len(args))
		}

		var total int
		countSQL := "SELECT COUNT(*) FROM contacts c WHERE " + whereExpr
		if err := tx.QueryRow(r.Context(), countSQL, args...).Scan(&total); err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "count error", ErrCodeDatabase)
			return
		}

		orderClause := "c.created_at DESC"
		if sortParam == "score" {
			orderClause = "c.lead_score DESC, c.created_at DESC"
		}

		dataArgs := append(append([]interface{}{}, args...), limit, offset)
		dataSQL := fmt.Sprintf(
			`SELECT c.id, c.agent_id, c.first_name, c.last_name, c.email, c.phone, c.source,
			        c.folder_id, cf.name,
			        c.lead_score, c.lead_score_signals, c.previous_lead_score,
			        c.created_at, c.updated_at
			 FROM contacts c
			 LEFT JOIN contact_folders cf ON cf.id = c.folder_id
			 WHERE %s ORDER BY %s LIMIT $%d OFFSET $%d`,
			whereExpr, orderClause, len(dataArgs)-1, len(dataArgs),
		)

		rows, err := tx.Query(r.Context(), dataSQL, dataArgs...)
		if err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "query error", ErrCodeDatabase)
			return
		}
		defer rows.Close()

		contacts := make([]Contact, 0)
		for rows.Next() {
			var c Contact
			if err := rows.Scan(&c.ID, &c.AgentID, &c.FirstName, &c.LastName, &c.Email, &c.Phone, &c.Source, &c.FolderID, &c.FolderName, &c.LeadScore, &c.LeadScoreSignals, &c.PreviousLeadScore, &c.CreatedAt, &c.UpdatedAt); err != nil {
				respondErrorWithCode(w, http.StatusInternalServerError, "scan error", ErrCodeDatabase)
				return
			}
			contacts = append(contacts, c)
		}

		if err := tx.Commit(r.Context()); err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "commit error", ErrCodeDatabase)
			return
		}

		respondJSON(w, http.StatusOK, map[string]interface{}{
			"contacts": contacts,
			"total":    total,
		})
	}
}

func GetContact(pool *pgxpool.Pool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		agentID := middleware.AgentUUIDFromContext(r.Context())
		id := chi.URLParam(r, "id")

		tx, err := database.BeginWithRLS(r.Context(), pool, agentID)
		if err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "database error", ErrCodeDatabase)
			return
		}
		defer tx.Rollback(r.Context())

		var c Contact
		err = tx.QueryRow(r.Context(),
			`SELECT c.id, c.agent_id, c.first_name, c.last_name, c.email, c.phone, c.source,
			        c.folder_id, cf.name,
			        c.lead_score, c.lead_score_signals, c.previous_lead_score,
			        c.created_at, c.updated_at
			 FROM contacts c
			 LEFT JOIN contact_folders cf ON cf.id = c.folder_id
			 WHERE c.id = $1`,
			id,
		).Scan(&c.ID, &c.AgentID, &c.FirstName, &c.LastName, &c.Email, &c.Phone, &c.Source, &c.FolderID, &c.FolderName, &c.LeadScore, &c.LeadScoreSignals, &c.PreviousLeadScore, &c.CreatedAt, &c.UpdatedAt)
		if err != nil {
			respondErrorWithCode(w, http.StatusNotFound, "contact not found", ErrCodeNotFound)
			return
		}

		tx.Commit(r.Context())
		respondJSON(w, http.StatusOK, c)
	}
}

func CreateContact(pool *pgxpool.Pool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		agentID := middleware.AgentUUIDFromContext(r.Context())

		var body struct {
			FirstName string  `json:"first_name"`
			LastName  string  `json:"last_name"`
			Email     *string `json:"email"`
			Phone     *string `json:"phone"`
			Source    *string `json:"source"`
			FolderID  *string `json:"folder_id"`
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			respondErrorWithCode(w, http.StatusBadRequest, "invalid JSON", ErrCodeBadRequest)
			return
		}
		if body.FirstName == "" || body.LastName == "" {
			respondErrorWithCode(w, http.StatusBadRequest, "first_name and last_name are required", ErrCodeBadRequest)
			return
		}
		// Input validation
		for _, v := range []struct {
			f, val string
			max    int
		}{
			{"first_name", body.FirstName, 100},
			{"last_name", body.LastName, 100},
		} {
			if err := validateMaxLen(v.f, v.val, v.max); err != nil {
				respondErrorWithCode(w, http.StatusBadRequest, err.Error(), ErrCodeBadRequest)
				return
			}
		}
		if body.Email != nil {
			if err := validateEmail(*body.Email); err != nil {
				respondErrorWithCode(w, http.StatusBadRequest, err.Error(), ErrCodeBadRequest)
				return
			}
		}
		if body.Phone != nil {
			if err := validateMaxLen("phone", *body.Phone, 20); err != nil {
				respondErrorWithCode(w, http.StatusBadRequest, err.Error(), ErrCodeBadRequest)
				return
			}
		}
		if body.Source != nil {
			if err := validateMaxLen("source", *body.Source, 50); err != nil {
				respondErrorWithCode(w, http.StatusBadRequest, err.Error(), ErrCodeBadRequest)
				return
			}
		}

		tx, err := database.BeginWithRLS(r.Context(), pool, agentID)
		if err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "database error", ErrCodeDatabase)
			return
		}
		defer tx.Rollback(r.Context())

		var c Contact
		err = tx.QueryRow(r.Context(),
			`INSERT INTO contacts (agent_id, first_name, last_name, email, phone, source, folder_id)
			 VALUES ($1, $2, $3, $4, $5, $6, $7)
			 RETURNING id, agent_id, first_name, last_name, email, phone, source, folder_id, NULL::text,
			           lead_score, lead_score_signals, previous_lead_score, created_at, updated_at`,
			agentID, body.FirstName, body.LastName, body.Email, body.Phone, body.Source, body.FolderID,
		).Scan(&c.ID, &c.AgentID, &c.FirstName, &c.LastName, &c.Email, &c.Phone, &c.Source, &c.FolderID, &c.FolderName, &c.LeadScore, &c.LeadScoreSignals, &c.PreviousLeadScore, &c.CreatedAt, &c.UpdatedAt)
		if err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "create failed", ErrCodeDatabase)
			return
		}

		// Auto-create a deal in "Lead" stage for every new contact
		_, _ = tx.Exec(r.Context(),
			`INSERT INTO deals (contact_id, agent_id, stage_id, title)
			 SELECT $1, $2, id, $3
			 FROM deal_stages WHERE LOWER(name) = 'lead' LIMIT 1`,
			c.ID, agentID, body.FirstName+" "+body.LastName,
		)

		// Auto-create a document folder for the contact
		_, _ = tx.Exec(r.Context(),
			`INSERT INTO document_folders (agent_id, name, contact_id) VALUES ($1, $2, $3)
			 ON CONFLICT DO NOTHING`,
			agentID, body.FirstName+" "+body.LastName, c.ID,
		)

		// Compute lead score for the new contact
		if err := scoring.ComputeLeadScore(r.Context(), tx, c.ID); err != nil {
			log.Printf("scoring: ComputeLeadScore failed for contact %s: %v", c.ID, err)
		}

		tx.Commit(r.Context())
		respondJSON(w, http.StatusCreated, c)
	}
}

func UpdateContact(pool *pgxpool.Pool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		agentID := middleware.AgentUUIDFromContext(r.Context())
		id := chi.URLParam(r, "id")

		var body map[string]interface{}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			respondErrorWithCode(w, http.StatusBadRequest, "invalid JSON", ErrCodeBadRequest)
			return
		}

		tx, err := database.BeginWithRLS(r.Context(), pool, agentID)
		if err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "database error", ErrCodeDatabase)
			return
		}
		defer tx.Rollback(r.Context())

		// Build partial update
		setClauses := "updated_at = NOW()"
		args := []interface{}{}
		allowed := []string{"first_name", "last_name", "email", "phone", "source", "folder_id"}
		for _, field := range allowed {
			if val, ok := body[field]; ok {
				args = append(args, val)
				setClauses += fmt.Sprintf(", %s = $%d", field, len(args))
			}
		}
		args = append(args, id)

		var c Contact
		err = tx.QueryRow(r.Context(),
			fmt.Sprintf(`UPDATE contacts SET %s WHERE id = $%d
			 RETURNING id, agent_id, first_name, last_name, email, phone, source, folder_id,
			 (SELECT name FROM contact_folders WHERE id = contacts.folder_id),
			 lead_score, lead_score_signals, previous_lead_score,
			 created_at, updated_at`, setClauses, len(args)),
			args...,
		).Scan(&c.ID, &c.AgentID, &c.FirstName, &c.LastName, &c.Email, &c.Phone, &c.Source, &c.FolderID, &c.FolderName, &c.LeadScore, &c.LeadScoreSignals, &c.PreviousLeadScore, &c.CreatedAt, &c.UpdatedAt)
		if err != nil {
			respondErrorWithCode(w, http.StatusNotFound, "contact not found", ErrCodeNotFound)
			return
		}

		// Recompute lead score after contact update
		if err := scoring.ComputeLeadScore(r.Context(), tx, c.ID); err != nil {
			log.Printf("scoring: ComputeLeadScore failed for contact %s: %v", c.ID, err)
		}

		tx.Commit(r.Context())
		respondJSON(w, http.StatusOK, c)
	}
}

func DeleteContact(pool *pgxpool.Pool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		agentID := middleware.AgentUUIDFromContext(r.Context())
		id := chi.URLParam(r, "id")

		tx, err := database.BeginWithRLS(r.Context(), pool, agentID)
		if err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "database error", ErrCodeDatabase)
			return
		}
		defer tx.Rollback(r.Context())

		result, err := tx.Exec(r.Context(), `DELETE FROM contacts WHERE id = $1`, id)
		if err != nil || result.RowsAffected() == 0 {
			respondErrorWithCode(w, http.StatusNotFound, "contact not found", ErrCodeNotFound)
			return
		}

		tx.Commit(r.Context())
		w.WriteHeader(http.StatusNoContent)
	}
}

func GoingColdCount(pool *pgxpool.Pool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		agentID := middleware.AgentUUIDFromContext(r.Context())
		if agentID == "" {
			respondErrorWithCode(w, http.StatusUnauthorized, "unauthorized", ErrCodeUnauthorized)
			return
		}

		tx, err := database.BeginWithRLS(r.Context(), pool, agentID)
		if err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "database error", ErrCodeDatabase)
			return
		}
		defer tx.Rollback(r.Context())

		var count int
		err = tx.QueryRow(r.Context(),
			`SELECT COUNT(*) FROM contacts c
			 WHERE c.lead_score < 20
			 AND NOT EXISTS (
			     SELECT 1 FROM activities a
			     WHERE a.contact_id = c.id
			     AND a.created_at > NOW() - INTERVAL '14 days'
			 )`).Scan(&count)
		if err != nil {
			respondErrorWithCode(w, http.StatusInternalServerError, "query error", ErrCodeDatabase)
			return
		}

		tx.Commit(r.Context())
		respondJSON(w, http.StatusOK, map[string]int{"count": count})
	}
}

File: ai-service/app/tools/contacts.py
"""
Contact tools: search, view, create, update, delete contacts and buyer profiles.
"""
import logging

import httpx
import psycopg2.extras

from app.config import BACKEND_URL, AI_SERVICE_SECRET
from app.database import get_conn, run_query

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _q(fn):
    """Open a connection + RealDictCursor, pass cursor to fn, return result."""
    with get_conn() as conn:
        cur = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
        return fn(cur)


def _dynamic_update(cur, table: str, allowed: set, id_col: str, id_val, fields: dict):
    """Build and execute a dynamic UPDATE. Returns the updated row id or None."""
    clean = {k: v for k, v in fields.items() if k in allowed}
    if not clean:
        return None
    set_clause = ", ".join(f"{k} = %s" for k in clean)
    vals = list(clean.values()) + [id_val]
    cur.execute(
        f"UPDATE {table} SET {set_clause}, updated_at = NOW() "
        f"WHERE {id_col} = %s RETURNING id",
        vals,
    )
    return cur.fetchone()


# ---------------------------------------------------------------------------
# Tool definitions (Anthropic tool schema)
# ---------------------------------------------------------------------------

DEFINITIONS = [
    {
        "name": "search_contacts",
        "description": (
            "Search contacts by name, email, or lead source. "
            "Matches against first_name, last_name, email, and full name. "
            "Pass a full name like 'Rohan Batre' or partial like 'Rohan'. "
            "For 'my last contact' or 'most recent', omit query and set limit=1 — results sort newest first. "
            "Always call this before any operation needing a contact_id."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "query": {"type": "string", "description": "Name or email to search for"},
                "source": {"type": "string", "description": "Filter by lead source (e.g. Zillow, Referral)"},
                "limit": {"type": "integer", "description": "Max results to return (default 10)"},
            },
            "required": [],
        },
    },
    {
        "name": "get_contact_details",
        "description": (
            "Get full details for one contact including buyer profile if it exists. "
            "Use after search_contacts to drill into a specific person."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "contact_id": {"type": "string", "description": "UUID of the contact"},
            },
            "required": ["contact_id"],
        },
    },
    {
        "name": "get_contact_activities",
        "description": (
            "Get recent activity history (calls, emails, notes, showings, tasks) for a contact. "
            "Sorted newest first."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "contact_id": {"type": "string", "description": "UUID of the contact"},
                "limit": {"type": "integer", "description": "Max results (default 20)"},
            },
            "required": ["contact_id"],
        },
    },
    {
        "name": "get_buyer_profile",
        "description": (
            "Get buyer preferences for a contact: budget, beds/baths, locations, "
            "must-haves, deal-breakers, pre-approval status, timeline."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "contact_id": {"type": "string", "description": "UUID of the contact"},
            },
            "required": ["contact_id"],
        },
    },
    {
        "name": "create_contact",
        "description": (
            "Create a NEW contact (not in CRM yet) and auto-create a Lead deal. "
            "Requires first_name and last_name at minimum. "
            "WARNING: Do NOT use this to update existing contacts — use update_contact instead. "
            "If a contact with the same name already exists, this will return the existing contact."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "first_name": {"type": "string"},
                "last_name": {"type": "string"},
                "email": {"type": "string"},
                "phone": {"type": "string"},
                "source": {"type": "string", "description": "Lead source (e.g. Zillow, Referral, Cold Call, Open House)"},
            },
            "required": ["first_name", "last_name"],
        },
    },
    {
        "name": "update_contact",
        "description": "Update a contact's name, email, phone, or source.",
        "input_schema": {
            "type": "object",
            "properties": {
                "contact_id": {"type": "string"},
                "first_name": {"type": "string"},
                "last_name": {"type": "string"},
                "email": {"type": "string"},
                "phone": {"type": "string"},
                "source": {"type": "string"},
            },
            "required": ["contact_id"],
        },
    },
    {
        "name": "delete_contact",
        "description": (
            "Permanently delete a contact and ALL associated data "
            "(deals, activities, buyer profile, conversations). Cannot be undone."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "contact_id": {"type": "string", "description": "UUID of the contact to delete"},
            },
            "required": ["contact_id"],
        },
    },
    {
        "name": "create_buyer_profile",
        "description": (
            "Create a buyer profile with property preferences for a contact. "
            "Only one profile per contact — use update_buyer_profile if one exists."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "contact_id": {"type": "string"},
                "budget_min": {"type": "number", "description": "Minimum budget in dollars"},
                "budget_max": {"type": "number", "description": "Maximum budget in dollars"},
                "bedrooms": {"type": "integer"},
                "bathrooms": {"type": "number"},
                "locations": {"type": "array", "items": {"type": "string"}, "description": "Preferred areas/neighborhoods"},
                "must_haves": {"type": "array", "items": {"type": "string"}, "description": "Required features (e.g. pool, garage)"},
                "deal_breakers": {"type": "array", "items": {"type": "string"}, "description": "Unwanted features (e.g. HOA, busy street)"},
                "property_type": {"type": "string", "description": "e.g. Single Family, Condo, Townhouse"},
                "pre_approved": {"type": "boolean"},
                "pre_approval_amount": {"type": "number"},
                "timeline": {"type": "string", "description": "e.g. ASAP, 3 months, 6 months"},
                "notes": {"type": "string"},
            },
            "required": ["contact_id"],
        },
    },
    {
        "name": "update_buyer_profile",
        "description": "Update an existing buyer profile. Only sends changed fields.",
        "input_schema": {
            "type": "object",
            "properties": {
                "contact_id": {"type": "string"},
                "budget_min": {"type": "number"},
                "budget_max": {"type": "number"},
                "bedrooms": {"type": "integer"},
                "bathrooms": {"type": "number"},
                "locations": {"type": "array", "items": {"type": "string"}},
                "must_haves": {"type": "array", "items": {"type": "string"}},
                "deal_breakers": {"type": "array", "items": {"type": "string"}},
                "property_type": {"type": "string"},
                "pre_approved": {"type": "boolean"},
                "pre_approval_amount": {"type": "number"},
                "timeline": {"type": "string"},
                "notes": {"type": "string"},
            },
            "required": ["contact_id"],
        },
    },
    {
        "name": "get_ai_profile",
        "description": (
            "Get the AI-generated summary profile for a contact. "
            "Includes personality insights, communication preferences, and engagement history summary. "
            "If no profile exists, suggests using regenerate_ai_profile to create one."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "contact_id": {"type": "string", "description": "UUID of the contact"},
            },
            "required": ["contact_id"],
        },
    },
    {
        "name": "regenerate_ai_profile",
        "description": (
            "Trigger AI profile regeneration for a contact. "
            "Analyzes the contact's activities, emails, deals, and buyer profile to create an updated summary. "
            "Use when a profile is missing or outdated."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "contact_id": {"type": "string", "description": "UUID of the contact"},
            },
            "required": ["contact_id"],
        },
    },
]

# ---------------------------------------------------------------------------
# Tool classification
# ---------------------------------------------------------------------------

READ = {"search_contacts", "get_contact_details", "get_contact_activities", "get_buyer_profile", "get_ai_profile"}
AUTO_EXECUTE: set[str] = {"regenerate_ai_profile"}
WRITE = {"create_contact", "update_contact", "delete_contact", "create_buyer_profile", "update_buyer_profile"}

# ---------------------------------------------------------------------------
# Field sets for dynamic updates
# ---------------------------------------------------------------------------

_CONTACT_FIELDS = {"first_name", "last_name", "email", "phone", "source"}
_BUYER_FIELDS = {
    "budget_min", "budget_max", "bedrooms", "bathrooms", "locations",
    "must_haves", "deal_breakers", "property_type", "pre_approved",
    "pre_approval_amount", "timeline", "notes",
}

# ---------------------------------------------------------------------------
# Read handlers
# ---------------------------------------------------------------------------

def _search_contacts(agent_id: str, inp: dict):
    query = inp.get("query", "")
    source = inp.get("source")
    limit = inp.get("limit", 10)

    def go(cur):
        clauses = ["c.agent_id = %s"]
        params: list = [agent_id]

        if query:
            like = f"%{query}%"
            clauses.append(
                "(c.first_name ILIKE %s OR c.last_name ILIKE %s "
                "OR c.email ILIKE %s OR (c.first_name || ' ' || c.last_name) ILIKE %s)"
            )
            params.extend([like, like, like, like])

        if source:
            clauses.append("c.source = %s")
            params.append(source)

        params.append(limit)
        cur.execute(f"""
            SELECT c.id, c.first_name, c.last_name, c.email, c.phone, c.source,
                   c.created_at, MAX(a.created_at) AS last_activity_at
            FROM contacts c
            LEFT JOIN activities a ON a.contact_id = c.id
            WHERE {' AND '.join(clauses)}
            GROUP BY c.id
            ORDER BY c.created_at DESC
            LIMIT %s
        """, params)
        return [dict(r) for r in cur.fetchall()]

    return _q(go)


def _get_contact_details(agent_id: str, inp: dict):
    contact_id = inp["contact_id"]

    def go(cur):
        cur.execute(
            "SELECT * FROM contacts WHERE id = %s AND agent_id = %s",
            (contact_id, agent_id),
        )
        contact = cur.fetchone()
        if not contact:
            return {"error": "Contact not found"}

        cur.execute("SELECT * FROM buyer_profiles WHERE contact_id = %s", (contact_id,))
        bp = cur.fetchone()

        result = dict(contact)
        result["buyer_profile"] = dict(bp) if bp else None
        return result

    return _q(go)


def _get_contact_activities(agent_id: str, inp: dict):
    contact_id = inp["contact_id"]
    limit = inp.get("limit", 20)

    def go(cur):
        cur.execute("""
            SELECT id, type, body, due_date, priority, completed_at, created_at
            FROM activities
            WHERE contact_id = %s AND agent_id = %s
            ORDER BY created_at DESC LIMIT %s
        """, (contact_id, agent_id, limit))
        return [dict(r) for r in cur.fetchall()]

    return _q(go)


def _get_buyer_profile(agent_id: str, inp: dict):
    contact_id = inp["contact_id"]

    def go(cur):
        cur.execute(
            "SELECT id FROM contacts WHERE id = %s AND agent_id = %s",
            (contact_id, agent_id),
        )
        if not cur.fetchone():
            return {"error": "Contact not found"}

        cur.execute("SELECT * FROM buyer_profiles WHERE contact_id = %s", (contact_id,))
        bp = cur.fetchone()
        if not bp:
            return {"error": "No buyer profile exists for this contact. Use create_buyer_profile to add one."}
        return dict(bp)

    return _q(go)


def _get_ai_profile(agent_id: str, inp: dict):
    contact_id = inp["contact_id"]

    def go(cur):
        cur.execute(
            "SELECT id FROM contacts WHERE id = %s AND agent_id = %s",
            (contact_id, agent_id),
        )
        if not cur.fetchone():
            return {"error": "Contact not found"}

        cur.execute("SELECT * FROM ai_profiles WHERE contact_id = %s", (contact_id,))
        profile = cur.fetchone()
        if not profile:
            return {"message": "No AI profile exists for this contact. Use regenerate_ai_profile to create one."}
        return dict(profile)

    return _q(go)


async def _regenerate_ai_profile(agent_id: str, inp: dict) -> dict:
    contact_id = inp["contact_id"]
    url = f"{BACKEND_URL}/api/contacts/{contact_id}/ai-profile/regenerate"
    headers = {
        "X-AI-Service-Secret": AI_SERVICE_SECRET,
        "X-Agent-ID": agent_id,
        "Content-Type": "application/json",
    }
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            resp = await client.post(url, headers=headers)

        if resp.status_code in (200, 201):
            return resp.json()

        try:
            detail = resp.json().get("error", resp.text)
        except Exception:
            detail = resp.text
        return {"error": f"Profile regeneration failed ({resp.status_code}): {detail}"}

    except httpx.ConnectError:
        return {"error": "Go backend not reachable."}
    except httpx.TimeoutException:
        return {"error": "Profile regeneration timed out — try again."}
    except Exception as exc:
        logger.error("regenerate_ai_profile proxy failed: %s", exc)
        return {"error": f"Profile regeneration failed: {exc}"}


# ---------------------------------------------------------------------------
# Write handlers
# ---------------------------------------------------------------------------

def _create_contact(agent_id: str, inp: dict):
    def go(cur):
        # Dedup check: reject if a contact with same first+last name already exists
        cur.execute(
            """SELECT id, first_name, last_name, email, phone, source
               FROM contacts
               WHERE agent_id = %s
                 AND LOWER(first_name) = LOWER(%s)
                 AND LOWER(last_name) = LOWER(%s)
               LIMIT 1""",
            (agent_id, inp["first_name"], inp["last_name"]),
        )
        existing = cur.fetchone()
        if existing:
            # Auto-convert to update: merge any new fields into the existing contact
            updates = {}
            for field in ("email", "phone", "source"):
                new_val = inp.get(field)
                if new_val and new_val != existing.get(field):
                    updates[field] = new_val

            if updates:
                set_clause = ", ".join(f"{k} = %s" for k in updates)
                vals = list(updates.values()) + [existing["id"], agent_id]
                cur.execute(
                    f"UPDATE contacts SET {set_clause}, updated_at = NOW() "
                    f"WHERE id = %s AND agent_id = %s RETURNING id, first_name, last_name, email, phone, source",
                    vals,
                )
                updated = dict(cur.fetchone())
                updated["merged"] = True
                updated["updated_fields"] = list(updates.keys())
                return updated

            # Nothing new to add — just return the existing contact
            result = dict(existing)
            result["already_exists"] = True
            return result

        cur.execute("""
            INSERT INTO contacts (agent_id, first_name, last_name, email, phone, source)
            VALUES (%s, %s, %s, %s, %s, %s)
            RETURNING id, first_name, last_name, email, phone, source, created_at
        """, (agent_id, inp["first_name"], inp["last_name"],
              inp.get("email"), inp.get("phone"), inp.get("source")))
        contact = dict(cur.fetchone())

        # Auto-create a Lead deal for pipeline tracking
        cur.execute("""
            INSERT INTO deals (contact_id, agent_id, stage_id, title)
            SELECT %s, %s, id, %s FROM deal_stages WHERE LOWER(name) = 'lead' LIMIT 1
            RETURNING id
        """, (contact["id"], agent_id, f"{inp['first_name']} {inp['last_name']}"))
        deal = cur.fetchone()
        if deal:
            contact["deal_id"] = deal["id"]
            contact["pipeline_stage"] = "Lead"

        return contact

    return _q(go)


def _update_contact(agent_id: str, inp: dict):
    inp = dict(inp)
    contact_id = inp.pop("contact_id")

    def go(cur):
        clean = {k: v for k, v in inp.items() if k in _CONTACT_FIELDS}
        if not clean:
            return {"error": "No valid fields to update"}
        set_clause = ", ".join(f"{k} = %s" for k in clean)
        vals = list(clean.values()) + [contact_id, agent_id]
        cur.execute(
            f"UPDATE contacts SET {set_clause}, updated_at = NOW() "
            f"WHERE id = %s AND agent_id = %s RETURNING id",
            vals,
        )
        row = cur.fetchone()
        if not row:
            return {"error": "Contact not found"}
        return {"updated": True, "contact_id": contact_id}

    return _q(go)


def _delete_contact(agent_id: str, inp: dict):
    contact_id = inp["contact_id"]

    def go(cur):
        # Verify ownership and get name for confirmation message
        cur.execute(
            "SELECT first_name, last_name FROM contacts WHERE id = %s AND agent_id = %s",
            (contact_id, agent_id),
        )
        contact = cur.fetchone()
        if not contact:
            return {"error": "Contact not found"}

        name = f"{contact['first_name']} {contact['last_name']}"

        # Single DELETE — ON DELETE CASCADE handles buyer_profiles, ai_profiles,
        # deals, activities, conversations, and conversations→messages
        cur.execute(
            "DELETE FROM contacts WHERE id = %s AND agent_id = %s",
            (contact_id, agent_id),
        )
        return {"deleted": True, "contact": name}

    return _q(go)


def _create_buyer_profile(agent_id: str, inp: dict):
    inp = dict(inp)
    contact_id = inp.pop("contact_id")

    def go(cur):
        # Verify contact ownership
        cur.execute(
            "SELECT id FROM contacts WHERE id = %s AND agent_id = %s",
            (contact_id, agent_id),
        )
        if not cur.fetchone():
            return {"error": "Contact not found"}

        # Check for existing profile
        cur.execute("SELECT id FROM buyer_profiles WHERE contact_id = %s", (contact_id,))
        if cur.fetchone():
            return {"error": "Buyer profile already exists. Use update_buyer_profile instead."}

        # Build columns/values from whatever was provided
        cols = ["contact_id"]
        vals = [contact_id]
        for field in _BUYER_FIELDS:
            if field in inp:
                cols.append(field)
                vals.append(inp[field])

        placeholders = ", ".join(["%s"] * len(vals))
        col_names = ", ".join(cols)
        cur.execute(
            f"INSERT INTO buyer_profiles ({col_names}) VALUES ({placeholders}) "
            f"RETURNING *",
            vals,
        )
        return dict(cur.fetchone())

    return _q(go)


def _update_buyer_profile(agent_id: str, inp: dict):
    inp = dict(inp)
    contact_id = inp.pop("contact_id")

    def go(cur):
        # Verify contact ownership
        cur.execute(
            "SELECT id FROM contacts WHERE id = %s AND agent_id = %s",
            (contact_id, agent_id),
        )
        if not cur.fetchone():
            return {"error": "Contact not found"}

        row = _dynamic_update(cur, "buyer_profiles", _BUYER_FIELDS, "contact_id", contact_id, inp)
        if row is None:
            if not {k for k in inp if k in _BUYER_FIELDS}:
                return {"error": "No valid fields to update"}
            return {"error": "No buyer profile exists. Use create_buyer_profile first."}
        return {"updated": True, "contact_id": contact_id}

    return _q(go)


# ---------------------------------------------------------------------------
# Dispatchers (dict-based, no if/elif chains)
# ---------------------------------------------------------------------------

_READ_DISPATCH = {
    "search_contacts": _search_contacts,
    "get_contact_details": _get_contact_details,
    "get_contact_activities": _get_contact_activities,
    "get_buyer_profile": _get_buyer_profile,
    "get_ai_profile": _get_ai_profile,
}

_AUTO_DISPATCH: dict = {
    "regenerate_ai_profile": lambda aid, inp: _regenerate_ai_profile(aid, inp),
}

_WRITE_DISPATCH = {
    "create_contact": _create_contact,
    "update_contact": _update_contact,
    "delete_contact": _delete_contact,
    "create_buyer_profile": _create_buyer_profile,
    "update_buyer_profile": _update_buyer_profile,
}


async def execute(tool_name: str, tool_input: dict, agent_id: str) -> dict:
    """Execute a read tool."""
    handler = _READ_DISPATCH.get(tool_name)
    if not handler:
        raise ValueError(f"Unknown contact read tool: {tool_name}")
    return await run_query(lambda: handler(agent_id, tool_input))


async def execute_auto(tool_name: str, tool_input: dict, agent_id: str) -> dict:
    """Execute an auto-execute tool."""
    handler = _AUTO_DISPATCH.get(tool_name)
    if not handler:
        raise ValueError(f"Unknown contact auto tool: {tool_name}")
    return await handler(agent_id, tool_input)


async def execute_write(tool_name: str, tool_input: dict, agent_id: str) -> dict:
    """Execute a confirmed write tool."""
    handler = _WRITE_DISPATCH.get(tool_name)
    if not handler:
        raise ValueError(f"Unknown contact write tool: {tool_name}")
    return await run_query(lambda: handler(agent_id, tool_input))
